classes/migration/upgrade/v3_4_0/PreflightCheckMigration.php

```diff
@@ -31,745 +34,1137 @@ abstract protected function getContextTable(): string;
-            }
-            // Clean orphaned edit_decisions entries by editor_id
-            $orphanedIds = DB::table('edit_decisions AS ed')->leftJoin('users AS u', 'u.user_id', '=', 'ed.editor_id')->whereNull('u.user_id')->distinct()->pluck('ed.editor_id');
-            foreach ($orphanedIds as $editorId) {
-                $this->_installer->log("Removing orphaned edit_decisions entry for missing editor_id {$editorId}");
-                DB::table('edit_decisions')->where('editor_id', '=', $editorId)->delete();
-            }
-            // Clean orphaned edit_decisions entries by submission_id
-            $orphanedIds = DB::table('edit_decisions AS ed')->leftJoin('submissions AS s', 's.submission_id', '=', 'ed.submission_id')->whereNull('s.submission_id')->distinct()->pluck('ed.submission_id');
-            foreach ($orphanedIds as $submissionId) {
-                $this->_installer->log("Removing orphaned edit_decisions entries for missing submission_id {$submissionId}");
-                DB::table('edit_decisions')->where('submission_id', '=', $submissionId)->delete();
-            }
-            // Clean orphaned submission_comments entries by submission_id
-            $orphanedIds = DB::table('submission_comments AS sc')->leftJoin('submissions AS s', 's.submission_id', '=', 'sc.submission_id')->whereNull('s.submission_id')->distinct()->pluck('sc.submission_id');
-            foreach ($orphanedIds as $submissionId) {
-                $this->_installer->log("Removing orphaned submission_comments entry for missing submission_id {$submissionId}");
-                DB::table('submission_comments')->where('submission_id', '=', $submissionId)->delete();
-            }
+        $this->addTableProcessor('citations', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~1 entities: publication_id->publications.publication_id
+            $affectedRows += $this->deleteRequiredReference('citations', 'publication_id', 'publications', 'publication_id');
+            return $affectedRows;
+        });
 
-            // Clean orphaned submission_comments entries by author_id
-            $orphanedIds = DB::table('submission_comments AS sc')->leftJoin('users AS u', 'u.user_id', '=', 'sc.author_id')->whereNull('u.user_id')->distinct()->pluck('sc.author_id');
-            foreach ($orphanedIds as $userId) {
-                $this->_installer->log("Removing orphaned submission_comments entry for missing author_id {$userId}");
-                DB::table('submission_comments')->where('author_id', '=', $userId)->delete();
-            }
+        $this->addTableProcessor('email_log', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~1 entities: sender_id->users.user_id
+            $affectedRows += $this->deleteRequiredReference('email_log', 'sender_id', 'users', 'user_id', $this->ignoreZero('sender_id'));
```

Comment:
The `email_log.sender_id` relationship does not have a foreign key constraint (that's filed for future work in https://github.com/pkp/pkp-lib/issues/8333), and we should not be removing email log entries when a user account is removed. If anything, the missing entries (when #8333 is implemented) should be set to `null` and the email log should be otherwise preserved. Removing a user shouldn't destroy evidence of emails sent on a submission.

Reply:
I generally agree, I just don't know about the GDPR stuff (right to be forgotten) 🤔